src/utils/pgwire/src/pg_extended.rs

```diff
@@ -78,28 +78,44 @@ impl PgStatement {
         &self,
         portal_name: String,
         params: &[Bytes],
-        result_format: bool,
-        param_format: bool,
+        result_format_codes: Vec<bool>,
+        param_format_codes: Vec<bool>,
     ) -> PsqlResult<PgPortal<VS>>
     where
         VS: Stream<Item = RowSetResult> + Unpin + Send,
     {
-        let instance_query_string = self.prepared_statement.instance(params, param_format)?;
-
-        let row_description: Vec<PgFieldDescriptor> = if result_format {
+        let instance_query_string = self
+            .prepared_statement
+            .instance(params, &param_format_codes)?;
+
+        // Invariant check
+        if !result_format_codes.is_empty()
+            && result_format_codes.len() != self.row_description.len()
+            && result_format_codes.len() != 1
+        {
+            return Err(PsqlError::Internal(anyhow!(
+                "format codes length {} is not 0, 1 or equal to column length {}",
+                result_format_codes.len(),
+                self.row_description.len()
+            )));
+        }
+        let row_description: Vec<PgFieldDescriptor> = {
             let mut row_description = self.row_description.clone();
+            let default_format = result_format_codes.first().copied().unwrap_or(false);
+            let mut format_iter = result_format_codes.iter();
+            row_description.iter_mut().for_each(|desc| {
+                let result_format = format_iter.next().copied().unwrap_or(default_format);
```

Comment:
Is it possible to wrap this checking and iterating logic in a helper, so that we do not need to repeat them at 3 places?

Reply:
> Is it possible to wrap this checking and iterating logic in a helper, so that we do not need to repeat them at 3 places?

Maybe we can implement a specific iterator for format code.🤔

Reply:
Agree.